src/elm/Pages/Build/Logs.elm

```diff
@@ -158,6 +158,69 @@ getLog resource get logs =
         |> List.head
 
 
+{-| addLog : takes incoming log and logs and adds log when not present
+-}
+addLog : Log -> Logs -> Logs
+addLog incomingLog logs =
+    RemoteData.succeed
+        (safeDecodeLogData incomingLog)
+        :: logs
+
+
+{-| updateLog : takes incoming log and logs and updates the appropriate log data
+-}
+updateLog : Log -> Logs -> Logs
+updateLog incomingLog logs =
+    updateIf
+        (\log ->
+            case log of
+                RemoteData.Success log_ ->
+                    incomingLog.id == log_.id && incomingLog.rawData /= log_.rawData
+
+                _ ->
+                    True
+        )
+        (\_ ->
+            RemoteData.succeed <|
+                safeDecodeLogData incomingLog
+        )
+        logs
+
+
+{-| safeDecodeLogData : takes log and decodes the data if it exists and does not exceed the size limit.
+-}
+safeDecodeLogData : Log -> Log
+safeDecodeLogData log =
+    { log
+        | decodedLogs =
+            if isEmpty log then
+                logEmptyMessage
+
+            else if exceedsSizeLimit log then
+                logSizeExceededMessage
+
+            else
+                Util.base64Decode log.rawData
+    }
+
+
+{-| logEmptyMessage : returns the default message when log data does not exist.
+-}
+logEmptyMessage : String
+logEmptyMessage =
+    "The build has not written logs to this step yet."
+
+
+{-| logSizeExceededMessage : returns the default message when a log exceeds the size limit.
+-}
+logSizeExceededMessage : String
+logSizeExceededMessage =
+    "The data for this log exceeds the size limit ("
+        ++ Util.formatFilesize logSizeLimit
+        ++ ").\n"
+        ++ "To view this log use the CLI or click the 'download' link in the top right corner of this step (downloading may take a few moments, depending on the size of the file)."
```

Comment:
this message could definitely be tweaked. 
what would provide the most value to the user?